File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/domain/ProjectsClient.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// domain/ProjectsClient.kt
//
// Domain client for projects operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import com.grey.sdk.grpc.PaginationResponse
import com.grey.sdk.grpc.ProjectResponse
import com.grey.sdk.grpc.ProjectsGrpcService
import com.grey.sdk.grpc.ProjectsListResponse

/**
 * Project data exposed to consumers.
 */
data class Project(
    val id: String,
    val name: String,
    val description: String?,
    val ownerId: String,
    val createdAt: Long,
    val updatedAt: Long
)

/**
 * Pagination info exposed to consumers.
 */
data class Pagination(
    val page: Int,
    val pageSize: Int,
    val totalPages: Int,
    val totalItems: Int
)

/**
 * Projects list data exposed to consumers.
 */
data class ProjectsData(
    val projects: List<Project>,
    val pagination: Pagination
)

/**
 * Domain client for projects operations.
 * All methods are suspend functions that return GreyResult.
 *
 * @property grpcService The underlying gRPC service
 * @property authClient The auth client for obtaining access tokens
 */
class ProjectsClient internal constructor(
    private val grpcService: ProjectsGrpcService,
    private val authClient: AuthClient
) {
    /**
     * List projects with pagination.
     * Requires authentication.
     *
     * @param page Page number (1-indexed, default: 1)
     * @param pageSize Items per page (default: 20)
     * @return GreyResult containing ProjectsData or error
     */
    suspend fun listProjects(page: Int = 1, pageSize: Int = 20): GreyResult<ProjectsData> {
        val accessToken = authClient.accessToken
            ?: return GreyResult.failure(GreyError.unauthenticated())

        if (page < 1) {
            return GreyResult.failure(GreyError.invalidArgument("Page must be >= 1"))
        }
        if (pageSize < 1 || pageSize > 100) {
            return GreyResult.failure(GreyError.invalidArgument("Page size must be between 1 and 100"))
        }

        return grpcService.listProjects(accessToken, page, pageSize)
            .map { response -> response.toProjectsData() }
    }

    /**
     * Create a new project.
     * Requires authentication.
     *
     * @param name Project name
     * @param description Optional project description
     * @return GreyResult containing created Project or error
     */
    suspend fun createProject(name: String, description: String? = null): GreyResult<Project> {
        val accessToken = authClient.accessToken
            ?: return GreyResult.failure(GreyError.unauthenticated())

        if (name.isBlank()) {
            return GreyResult.failure(GreyError.invalidArgument("Project name is required"))
        }

        return grpcService.createProject(accessToken, name, description)
            .map { response -> response.toProject() }
    }

    /**
     * Extension to convert gRPC response to domain model.
     */
    private fun ProjectsListResponse.toProjectsData(): ProjectsData = ProjectsData(
        projects = projects.map { it.toProject() },
        pagination = pagination.toPagination()
    )

    /**
     * Extension to convert gRPC response to domain model.
     */
    private fun ProjectResponse.toProject(): Project = Project(
        id = id,
        name = name,
        description = description,
        ownerId = ownerId,
        createdAt = createdAt,
        updatedAt = updatedAt
    )

    /**
     * Extension to convert gRPC response to domain model.
     */
    private fun PaginationResponse.toPagination(): Pagination = Pagination(
        page = page,
        pageSize = pageSize,
        totalPages = totalPages,
        totalItems = totalItems
    )
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/Domain/ProjectsClient.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// Domain/ProjectsClient.swift
//
// Domain client for projects operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

import Foundation

// MARK: - Domain Models

/// Project data exposed to consumers.
public struct Project: Sendable {
    public let id: String
    public let name: String
    public let description: String?
    public let ownerId: String
    public let createdAt: Int64
    public let updatedAt: Int64
    
    public init(
        id: String,
        name: String,
        description: String?,
        ownerId: String,
        createdAt: Int64,
        updatedAt: Int64
    ) {
        self.id = id
        self.name = name
        self.description = description
        self.ownerId = ownerId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

/// Pagination info exposed to consumers.
public struct Pagination: Sendable {
    public let page: Int
    public let pageSize: Int
    public let totalPages: Int
    public let totalItems: Int
    
    public init(page: Int, pageSize: Int, totalPages: Int, totalItems: Int) {
        self.page = page
        self.pageSize = pageSize
        self.totalPages = totalPages
        self.totalItems = totalItems
    }
}

/// Projects list data exposed to consumers.
public struct ProjectsData: Sendable {
    public let projects: [Project]
    public let pagination: Pagination
    
    public init(projects: [Project], pagination: Pagination) {
        self.projects = projects
        self.pagination = pagination
    }
}

// MARK: - Client

/// Domain client for projects operations.
/// All methods are async functions that throw GreyError on failure.
public final class ProjectsClient: Sendable {
    
    private let grpcService: ProjectsGRPCService
    private let authClient: AuthClient
    
    // MARK: - Initialization
    
    internal init(grpcService: ProjectsGRPCService, authClient: AuthClient) {
        self.grpcService = grpcService
        self.authClient = authClient
    }
    
    // MARK: - Operations
    
    /// List projects with pagination.
    /// Requires authentication.
    /// - Parameters:
    ///   - page: Page number (1-indexed, default: 1)
    ///   - pageSize: Items per page (default: 20)
    /// - Returns: ProjectsData on success
    /// - Throws: GreyError on failure
    public func listProjects(page: Int = 1, pageSize: Int = 20) async throws -> ProjectsData {
        guard let accessToken = authClient.accessToken else {
            throw GreyError.unauthenticated()
        }
        
        guard page >= 1 else {
            throw GreyError.invalidArgument("Page must be >= 1")
        }
        
        guard pageSize >= 1 && pageSize <= 100 else {
            throw GreyError.invalidArgument("Page size must be between 1 and 100")
        }
        
        let response = try await grpcService.listProjects(
            accessToken: accessToken,
            page: page,
            pageSize: pageSize
        )
        
        return ProjectsData(
            projects: response.projects.map { project in
                Project(
                    id: project.id,
                    name: project.name,
                    description: project.description,
                    ownerId: project.ownerId,
                    createdAt: project.createdAt,
                    updatedAt: project.updatedAt
                )
            },
            pagination: Pagination(
                page: response.pagination.page,
                pageSize: response.pagination.pageSize,
                totalPages: response.pagination.totalPages,
                totalItems: response.pagination.totalItems
            )
        )
    }
    
    /// Create a new project.
    /// Requires authentication.
    /// - Parameters:
    ///   - name: Project name
    ///   - description: Optional project description
    /// - Returns: Created Project on success
    /// - Throws: GreyError on failure
    public func createProject(name: String, description: String? = nil) async throws -> Project {
        guard let accessToken = authClient.accessToken else {
            throw GreyError.unauthenticated()
        }
        
        guard !name.isEmpty else {
            throw GreyError.invalidArgument("Project name is required")
        }
        
        let response = try await grpcService.createProject(
            accessToken: accessToken,
            name: name,
            description: description
        )
        
        return Project(
            id: response.id,
            name: response.name,
            description: response.description,
            ownerId: response.ownerId,
            createdAt: response.createdAt,
            updatedAt: response.updatedAt
        )
    }
}
